### PR TITLE
add queryAll SOQL operation

### PR DESCRIPTION
**Pull request by jster1357**
add queryAll SOQL operation

### CODE REVIEW

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java [hunk 1]
```diff
@@ -75,6 +76,11 @@
   @Macro
   private String offset;
 
+  @Name(SalesforceSourceConstants.PROPERTY_OPERATION)
+  @Description("If set to query, the query result will only return current rows. If set to queryAll, " +
+          "all records, including deletes will be sourced")
+  private String operation;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:83]: should this param be macro enabled?
Also, seems like we should support null operation for backward compatibility, so let's make it nullable and default to 'query' when operation is not specified

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java [hunk 2]
```diff
@@ -269,4 +276,17 @@ private int parseUnitValue(String propertyName, String value) {
   private ZonedDateTime parseDatetime(String datetime) throws DateTimeParseException {
     return StringUtils.isBlank(datetime) ? null : ZonedDateTime.parse(datetime, DateTimeFormatter.ISO_DATE_TIME);
   }
+
+  public String getOperation() {
+    return operation;
+  }
+
+  public OperationEnum getOperationEnum() {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java]: where is this called?

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java [hunk 3]
```diff
@@ -96,11 +97,13 @@
...
    */
-  private static BatchInfo[] runBulkQuery(BulkConnection bulkConnection, String query, boolean enablePKChunk)
+  private static BatchInfo[] runBulkQuery(BulkConnection bulkConnection, String query,
+                                          boolean enablePKChunk, String operation)
     throws AsyncApiException, IOException {
 
     SObjectDescriptor sObjectDescriptor = SObjectDescriptor.fromQuery(query);
-    JobInfo job = SalesforceBulkUtil.createJob(bulkConnection, sObjectDescriptor.getName(), OperationEnum.query, null);
+    JobInfo job = SalesforceBulkUtil.createJob(bulkConnection, sObjectDescriptor.getName(),
+            OperationEnum.valueOf(operation), null);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java]: handle null operation, annotate with @Nullable

**Comment by jster1357** [comment]
JIRA for the change: https://cdap.atlassian.net/browse/PLUGIN-701

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 1 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:83]: I have this set as a radio button, defaulted to "query"...does this work?

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 2 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java]: older code...I'll remove

File: src/test/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceSourceConfigBuilder.java [hunk 4]
```diff
@@ -123,9 +124,14 @@ public SalesforceSourceConfigBuilder setParent(String parent) {
     return this;
   }
 
+  public SalesforceSourceConfigBuilder setOperation(String operation) {
```
Comment (snehalakshmisha) [inline src/test/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceSourceConfigBuilder.java:127]: Where is this method called?

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 1 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:83]: That works when the user is building a pipeline in pipeline studio. 
But we also need to support the case where a pipeline JSON with an older version of salesforce plugin is imported. 
The old pipeline json won't have 'operation' config.

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java (hunk 3 above)
Comment (snehalakshmisha) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java]: You've defined SalesforceBaseSourceConfig.getOperationEnum() which implies `Operation.valueOf(enum)` can throw exceptions and that should probably be handled here as well

File: src/test/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceSourceConfigBuilder.java (hunk 4 above)
Comment (jster1357) [reply src/test/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceSourceConfigBuilder.java:127]: I haven't built any test cases for it. It was erroring out when I tried to build it so I added this get past the build.

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java [hunk 5]
```diff
@@ -269,4 +277,10 @@ private int parseUnitValue(String propertyName, String value) {
   private ZonedDateTime parseDatetime(String datetime) throws DateTimeParseException {
     return StringUtils.isBlank(datetime) ? null : ZonedDateTime.parse(datetime, DateTimeFormatter.ISO_DATE_TIME);
   }
+
+  public String getOperation() {
+    return operation == null ? "query" : operation;
```
Comment (snehalakshmisha) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java]: `query` should be a constant string

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java [hunk 6]
```diff
@@ -269,4 +277,10 @@ private int parseUnitValue(String propertyName, String value) {
   private ZonedDateTime parseDatetime(String datetime) throws DateTimeParseException {
     return StringUtils.isBlank(datetime) ? null : ZonedDateTime.parse(datetime, DateTimeFormatter.ISO_DATE_TIME);
   }
+
+  public String getOperation() {
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:298]: add code in validate() to check the operation is either query or queryAll. Add failure to failure collector if the operation is invalid

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 5 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java]: added a constant string

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java (hunk 3 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java]: added new method that deals with this

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 1 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:83]: got it. I've made it nullable

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 6 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:298]: so add similar logic to whats being done with validateFilters()?

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java [hunk 7]
```diff
@@ -15,6 +15,7 @@
  */
 package io.cdap.plugin.salesforce.plugin.source.batch;
 
+import com.sforce.async.OperationEnum;
```
Comment (snehalakshmisha) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:18]: If this isn't used in the class anymore it should be remvoed

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java [hunk 8]
```diff
@@ -75,6 +76,14 @@
   @Macro
   private String offset;
 
+  @Name(SalesforceSourceConstants.PROPERTY_OPERATION)
+  @Description("If set to query, the query result will only return current rows. If set to queryAll, " +
```
Comment (snehalakshmisha) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:80]: nit: Is there a way to highlight `query` and `queryAll` to indicate that it is a keyword?

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java [hunk 9]
```diff
@@ -75,6 +76,14 @@
...
+          "all records, including deletes will be sourced")
+  @Nullable
+  private String operation;
+
+  private static final String default_operation = "query";
```
Comment (snehalakshmisha) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java]: `default_operation` should be in upper case

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceMultiSourceConfig.java [hunk 10]
```diff
@@ -81,7 +81,8 @@ public SalesforceMultiSourceConfig(String referenceName,
                               @Nullable String whiteList,
                               @Nullable String blackList,
                               @Nullable String sObjectNameField,
-                              @Nullable String securityToken) {
+                              @Nullable String securityToken,
+                              @Nullable String operation) {
```
Comment (snehalakshmisha) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceMultiSourceConfig.java:85]: If we are passing operation in the constructor, we must want to use that value somewhere? Either set the value of the field here or pass it to super().

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java [hunk 11]
```diff
@@ -22,11 +22,7 @@
 import com.sforce.async.JobInfo;
 import com.sforce.async.JobStateEnum;
 import com.sforce.async.OperationEnum;
-import io.cdap.plugin.salesforce.BulkAPIBatchException;
-import io.cdap.plugin.salesforce.SObjectDescriptor;
-import io.cdap.plugin.salesforce.SalesforceBulkUtil;
-import io.cdap.plugin.salesforce.SalesforceConnectionUtil;
-import io.cdap.plugin.salesforce.SalesforceQueryUtil;
+import io.cdap.plugin.salesforce.*;
```
Comment (snehalakshmisha) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java]: We should not have * imports. Import only the classes you need.
You can do a `mvn clean package` after you address any comments for sanity check, that will catch issues some issues like this for you

Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java]: Thanks. I fixed this on my latest build.

Comment (snehalakshmisha) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/SalesforceSplitUtil.java]: You can turn off automatic wildcard imports - https://intellij-support.jetbrains.com/hc/en-us/community/posts/206203659-Turn-off-Wildcard-imports-

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 7 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:18]: done

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 9 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java]: done

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceMultiSourceConfig.java (hunk 10 above)
Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceMultiSourceConfig.java:85]: added it to super

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java (hunk 6 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:298]: yes. convert operation string to enum. If that fails (exception is thrown), then add a failure to the failure colllector.

Comment (jster1357) [reply src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:298]: I just pushed some code that does validation.

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java [hunk 12]
```diff
@@ -203,6 +218,16 @@ private void validateIntervalFilterProperty(String propertyName, String datetime
     }
   }
 
+  private void validateOperationProperty(String propertyName, String operation) {
+    try {
+      OperationEnum.valueOf(operation);
+    } catch (InvalidConfigException e) {
+      throw new InvalidConfigException(
+              String.format("Invalid Query Operation: '%s' value: '%s'. Valid values are query and queryAll.",
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java:226]: change the string to -
String.format("Invalid Query Operation: '%s'. Valid values are query and queryAll.", operation)

**Review by rmstar — APPROVED** [review]
please squash commits after addressing the minor comment

**Pull request merged**